unified_planning/io/anml_reader.py

```diff
@@ -0,0 +1,810 @@
+                        left_bound = duration_exp[0]
+                        right_bound = duration_exp[2]
+                        assert left_bound[0][0] == TK_DURATION
+                        assert right_bound[0][0] == TK_DURATION
+                        if left_bound[1] not in (TK_GT, TK_GE):
+                            left_bound, right_bound = right_bound, left_bound
+                        assert left_bound[1] in (TK_GT, TK_GE)
+                        assert right_bound[1] in (TK_LT, TK_LE)
+                        is_left_open = left_bound[1] == TK_GT
+                        is_right_open = right_bound[1] == TK_LT
+                        l_bound_exp = self._parse_expression(
+                            left_bound[2], action_parameters, types_map
+                        )
+                        r_bound_exp = self._parse_expression(
+                            right_bound[2], action_parameters, types_map
+                        )
+                        action.set_duration_constraint(
+                            DurationInterval(
+                                l_bound_exp, r_bound_exp, is_left_open, is_right_open
+                            )
+                        )
+            # end handle duration
+            elif c_assign + c_increase + c_decrease > 0:  # handle effects
+                up_timing, up_effect = self._parse_assignment(
+                    interval_and_exp, action_parameters, types_map
+                )
+                action._add_effect_instance(up_timing, up_effect)
+            # end handle effects
+            else:  # handle condition
+                assert c_when == 0, "when keyword used in an action precondition"
+                up_interval = self._parse_interval(interval_and_exp[0], types_map)
+                exp_res = interval_and_exp[1][0]
+                condition = self._parse_expression(
+                    exp_res, action_parameters, types_map
+                )
+                action.add_condition(up_interval, condition)
+
+    def _parse_parameters_def(
+        self, parameters_res: ParseResults, types_map: Dict[str, "Type"]
+    ) -> "OrderedDict[str, Type]":
+        up_params: "OrderedDict[str, Type]" = OrderedDict()
+        for parameter_res in parameters_res:
+            param_type_res = parameter_res[0]
+            param_name_res = parameter_res[1]
+            assert isinstance(param_name_res, str)
+            up_params[param_name_res] = self._parse_type_reference(
+                param_type_res, types_map
+            )
+        return up_params
+
+    def _parse_interval(
+        self,
+        interval_res: ParseResults,
+        types_map: Dict[str, "Type"],
+        is_global: bool = False,
+    ) -> Union["Timing", "TimeInterval"]:
+        if len(interval_res) == 0:
+            if is_global:
+                # TODO: decide/understand if an undefined interval outside of an action really means GlobalStartTiming or not
```

Comment:
Note this missing TODO

Reply:
It is not allowed to have expressions with an undefined interval, except for constants initialization (currently not supported). If it happens we should raise an exception.